packages/itwinui-react/src/core/Menu/MenuItem.tsx

```diff
@@ -115,16 +127,54 @@ export const MenuItem = React.forwardRef((props, forwardedRef) => {
     React.useState(false);
   const parent = React.useContext(MenuItemContext);
 
+  const dropdownMenuContext = React.useContext(DropdownMenuContext);
+
+  const nodeId = useFloatingNodeId();
+  const tree = useFloatingTree();
+  const parentId = useFloatingParentNodeId();
+
   const onVisibleChange = (open: boolean) => {
-    setIsSubmenuVisible(open);
+    if (open) {
+      // Once the menu is opened, reset focusOnSubmenu (since it is set to true when the right arrow is pressed)
+      setFocusOnSubmenu(false);
+
+      tree?.events.emit('submenuOpened', {
+        nodeId,
+        parentId,
+      } satisfies TreeEvent);
+    }
+
+    setIsSubmenuVisible(open || isNestedSubmenuVisible);
 
     // we don't want parent to close when mouse goes into a nested submenu,
     // so we need to let the parent know whether the submenu is still open.
     parent.setIsNestedSubmenuVisible(open);
   };
 
+  React.useEffect(() => {
+    const handleSubmenuOpened = (event: TreeEvent) => {
+      // Only one submenu in each menu can be open at a time
+      // So, if a sibling's submenu is opened, close this submenu
+      if (event.parentId === parentId && event.nodeId !== nodeId) {
+        setIsSubmenuVisible(false);
+        setIsNestedSubmenuVisible(false);
+      }
+    };
+
+    tree?.events.on('submenuOpened', handleSubmenuOpened);
+
+    return () => {
+      tree?.events.off('submenuOpened', handleSubmenuOpened);
+    };
+  }, [nodeId, parentId, tree?.events, tree?.nodesRef]);
+
   const popover = usePopover({
-    visible: isSubmenuVisible || isNestedSubmenuVisible,
+    nodeId,
+    visible:
+      isSubmenuVisible ||
+      isNestedSubmenuVisible ||
+      // to keep the submenu open when mouse enters it and then hovers out
+      dropdownMenuContext.lastHoveredNode?.parentId === nodeId,
```

Comment:
I don't think this is quite what we discussed. By "hover ≈ focus" I meant literally focusing an item (by calling `.focus()`) when it is hovered. Arrow-key navigation should continue from the last hovered item.

Reply:
> But triggering :focus-visible and then hovering over Item 4_2 does not do that.

If by "triggering `:focus-visible`" you mean in dev-tools, then it's usually not representative of real interactions. Usually you want to check both `:focus` and `:focus-visible`. And even then it is mainly useful for debugging styles (CSS), not behavior (JS). For proper debugging, you can use the same snippet mentioned in https://github.com/iTwin/iTwinUI/issues/1924#issuecomment-2014097457.

Nothing stands out to me in code, but if I had to guess, it's likely the `hover`+`focus` triggers in `usePopover` that's messing with it.

Reply:
The `setFocusOnSubmenu` logic could also be conflicting with it. Ideally we wouldn't need to use this hack at all.

Reply:
> By "hover ≈ focus" I meant literally focusing an item (by calling `.focus()`) when it is hovered

In [`4b29b10`](https://github.com/iTwin/iTwinUI/pull/1942/commits/4b29b1077625712e9997760558386df0b339f3d8), I removed the `useHover` trigger and but kept the `useFocus` trigger. Now, when we hover over a `MenuItem`, we call `.focus()` on itself inside the `onMouseEnter`. Since the `useFocus` trigger exists, hovering still indirectly opens a submenu.

Since "hover ≈ focus", one change is that earlier, opening a `DropdownMenu` where the first item has a submenu used to not open that submenu. But not since "hover ≈ focus", that submenu is opened. Not sure if this change in behavior is unideal.

We can see this when opening the leftmost menu button using a mouse click in the below recording:

<details>
<summary>Screen-recording</summary>

Video parts:
1. Second DropdownMenu mouse: no noticeable behavior change
2. First DropdownMenu mouse: `Item 1_3`'s submenu is not automatically open
3. Second DropdownMenu keyboard: no noticeable behavior change
4. First DropdownMenu keyboard: no noticeable behavior change

https://github.com/iTwin/iTwinUI/assets/45748283/aae38f59-907b-4f10-9946-f89209c03cf7

</details>

In the above-mentioned commit, I temporarily hard-coded [`visibleOnly`](https://floating-ui.com/docs/usefocus#visibleonly) to `false` in `Popover.tsx` to make this work. I will changing the hard coded value to be general if we decide to continue down this path.

> I had tried to call `.focus()` in `onMouseEnter` but kept getting unexpected behavior similar to this:

Like you suggested and I had alluded to before, the hover and focus triggers were for some reason not cooperating. The issue I mentioned is no longer present after removing the `onHover` trigger.

Reply:
The first submenu automatically opening when the main menu is opened is not acceptable. The user only meant to open the main menu. They should be expected to hover or arrow-right on the first item to open the submenu.

Ideally we'd be able to use [`focus({ focusVisible: true })`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus#focusvisible), but it has poor browser support.

So I think we'll need to look into the heuristics of how `visibleOnly` is implemented in floating-ui to figure out how we can emulate focusVisible.

Reply:
I noticed that maintaining an `isHovered` state variable (https://github.com/iTwin/iTwinUI/pull/1942/commits/5ac32c6643f3f23c93b5c7b12b15b3f92ad0d953) solves the issue of unexpectedly closing all submenus upon hovering over "Item 4_2" as mentioned in https://github.com/iTwin/iTwinUI/pull/1942#discussion_r1556174966.

In the same commit, I also brought back the hover trigger. This helped bring back the [`safePolygon`](https://floating-ui.com/docs/usehover#safepolygon). Everything still continues to work.

> The first submenu automatically opening when the main menu is opened is not acceptable.

I also was able to undo setting `visisbleOnly: false` in `useHover` in `Popover.tsx`. So, "Item_3" in the leftmost DropdownMenu now no longer opens its submenu upon opening the DropdownMenu after clicking the "Menu" button. It only opens upon hovering "Item 1_3".

<details>
<summary>Screen-recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/3c7b68d0-f355-4232-b7dd-f531b16d048e

</details>

Reply:
> > The first submenu automatically opening when the main menu is opened is not acceptable.
> 
> I also was able to undo setting `visisbleOnly: false` in `useHover` in `Popover.tsx`. So, "Item_3" in the leftmost DropdownMenu now no longer opens its submenu upon opening the DropdownMenu after clicking the "Menu" button. It only opens upon hovering "Item 1_3".

I think the root cause may not be fixed. Because the same problem can happen inside a nested menu when the first item has another submenu.

<details><summary>Screen recording</summary>

In here, the third level menu automatically opens when Item #4 is focused.

https://github.com/iTwin/iTwinUI/assets/9084735/fa6491e2-3ca4-449f-8b5f-37740e09fda4

</details> 

To avoid these complications, I'm now rethinking whether we should keep `focus` as the trigger.

Arrow keys are already used as the trigger, and in the recent UX call we also discussed using click/<kbd>Enter</kbd> as the trigger (#879). Together, these two triggers might make the `focus` trigger not as important.

If this change is too complicated, it can also happen in a separate follow-up PR.

Reply:
> In here, the third level menu automatically opens when `Item #4` is focused.

Ohh, I thought that was expected since we decided hover ≈ focus. As a result, since hovering over `Item #4` would open the third level submenu, focusing it also does the same.

> To avoid these complications, I'm now rethinking whether we should keep `focus` as the trigger.

Sounds good, this was my preference too because that is how the menus work in other UI libraries (e.g. [Radix UI](https://www.radix-ui.com/primitives/docs/components/dropdown-menu), [React Aria](https://react-spectrum.adobe.com/react-aria/Menu.html#submenus), [AriaKit](https://ariakit.org/examples/menu-nested)). Even the menu bar menus and context menus on my Mac work the same way (i.e. doesn't open the submenu on focus and only opens them using the <kbd>←</kbd>/<kbd>→</kbd> keys).

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/060fcda7-a378-495c-a205-5e7bb0c21d7a

</details>

I guess they might have been working differently (i.e. using focus as a trigger) on your Mac because of some change in your settings.

> If this change is too complicated, it can also happen in a separate follow-up PR.

Great, so, I will try working on removing the focus trigger and seeing how it works and also see whether a new PR would be better.

Reply:
> Ohh, I thought that was expected since we decided hover ≈ focus. As a result, since hovering over Item #4 would open the third level submenu, focusing it also does the same.

You can think of it as a one-directional relationship. Hovering should focus the item, but focusing by itself is its own separate thing that does not need to behave.

The more important thing to remember is the user's intent. Opening a menu automatically focuses the first menu-item, but this focus was not moved by the user. So they didn't intend to open the sub-menu. Compare this to a mouse user who needs to intentionally move their mouse from the trigger element into menu-item.

> Even the menu bar menus and context menus on my Mac work the same way … I guess they might have been working differently (i.e. using focus as a trigger) on your Mac because of some change in your settings.

I couldn't find any setting that I changed. It's very possible that the behavior was changed in macOS 14. I prefer it this way, but it very much needs to be done properly (which currently it is not).

> Great, so, I will try working on removing the focus trigger and seeing how it works and also see whether a new PR would be better.

If you intend to do a separate PR for the click trigger, then I would keep the focus trigger unchanged in this PR.

Reply:
> You can think of it as a one-directional relationship. Hovering should focus the item, but focusing by itself is its own separate thing that does not need to behave identically. 

Sounds good, with the recent changes, focusing a MenuItem does not open its submenu until <kbd>→</kbd> is pressed:

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/f2e8b202-b2fb-4678-8648-3e80420d4bcf

</details>

In contrast, hovering a MenuItem opens its submenu:

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/c0e1c690-a20a-4951-b7ba-0e7559e69255

</details>

> If you intend to do a separate PR for the click trigger, then I would keep the focus trigger unchanged in this PR.

Yes, I hope to work on the click trigger in a separate PR since that required more work than just adding the trigger.

I have currently removed the focus trigger to stop opening the submenu on focus, as mentioned in the point above. Is there any strong reason to keep the focus trigger in this PR instead of removing it?

Reply:
Adding back `focus` takes away some of the benefits from this PR. E.g. pressing <kbd>←</kbd> on a menu "X" will move focus to X's parent but still keep X open instead of closing it.

<details>
<summary>Screen recording</summary>

https://github.com/iTwin/iTwinUI/assets/45748283/05dffd03-bedf-4299-8cf8-fc633eed2094

</details>

> When both focus and click are absent, arrow keys are the only trigger, which is not sufficient.

To confirm, why are arrow keys and hover insufficient triggers?

Reply:
> To confirm, why are arrow keys and hover insufficient triggers?

With complex patterns like submenus, there should always be multiple affordances. A screen-reader or keyboard user might not even realize that they need to press the arrow keys to open submenus.

In an ideal scenario, we would have focus+click+arrow-keys as triggers (+hover for mouse users). The next best thing is click+arrow-keys (+hover).

("click" also implies <kbd>Enter</kbd>/<kbd>Space</kbd> keypress)

Reply:
Sounds good, added the click trigger manually instead of using `useClick` from floating UI (https://github.com/iTwin/iTwinUI/pull/1942/files/1b5f5f2f1f88375ebd8fe76b45887dc241d3ab54..a9d865f26b10f759df61210091fb0c0d00a16108).

So, the triggers now are hover, click (including <kbd>Enter</kbd>/<kbd>Space</kbd>), and arrow keys.